### ISSUE TITLE
Function template overloading 

### ISSUE THREAD

**Issue by Udayv56** [Udayv56/CPP#47]
We'll be discussing template function overloading using the code provided. Template function overloading is a powerful feature in C++ that allows us to define multiple functions with the same name but different template parameters or argument types.

In the given code, we have a class template `Uday` that contains a single data member `data` of type `T`. It also has a member function `display()` which is defined outside the class template.

```cpp
template <class T>
class Uday {
    public:
        T data;
        Uday(T a) {
            data = a;
        }
        void display();
};

template <class T>
void Uday<T>::display() {
    cout << data << endl;
}
```

Now, let's move on to the concept of template function overloading. In the code, we have two functions named `func`: one is a non-template function, and the other is a template function. Let's analyze them one by one.

```cpp
void func(int a) {
    cout << "I am non-template " << a << endl;
}

template <class T>
void func(T a) {
    cout << "I am template " << a << endl;
}
```

In this case, we have an overload of the function `func`. The first version of `func` takes an `int` parameter, and the second version is a template function that can accept any type `T`. It is important to note that the non-template function will be preferred over the template function if the exact match is found.

Now, let's examine the `main` function. We have several function calls to `func` with different arguments:

```cpp
func(4);     // exact match, highest priority
func(5.2);   // non-template function not an exact match, template function used
func('c');   // non-template function not an exact match, template function used
```

Here's what happens during the execution:

1. The first function call `func(4)` is an exact match to the non-template function `func(int a)`. Therefore, the non-template version is selected and executed, printing "I am non-template 4" to the console.

2. The second function call `func(5.2)` does not match the non-template function exactly. However, there is a template function that can accept a `double` argument, which is the closest match. Thus, the template function `func(T a)` is selected and executed, printing "I am template 5.2" to the console.

3. The third function call `func('c')` does not match the non-template function exactly. Similarly, there is a template function that can accept a `char` argument, so the template function `func(T a)` is again selected and executed, printing "I am template c" to the console.

In summary, template function overloading allows us to define functions with the same name but different parameter types or template parameters. During function calls, the compiler selects the most appropriate version based on the provided arguments. Non-template functions have higher priority over template functions if an exact match is found. If no exact match is found, the template function is used as a fallback option. This feature provides flexibility and reusability in our code.

I hope this explanation clarifies the concept of template function overloading for you. If you have any questions, please feel free to ask!

**OUTPUT-**

I am non template 4
I am template 5.2
I am template c
